### PR TITLE
Feat/challenge evm circuit

### PR DESCRIPTION
**Pull request by adria0**
The original PR was https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/990 but after renaming the branch to `feat/challenge-evm-circuit` github automatically closed it.

Depends on https://github.com/privacy-scaling-explorations/halo2/pull/115

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/execution.rs:595]:
```suggestion
        // height
```

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/util.rs:252]: I can't think about the reason to require phase 3 here? Would you like to explain it more?

**Review by han0110 — COMMENTED** [review]
First round pass. Overall looks good to me! Only left some small questions.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs:21]: Ah right make sense, just like the logic in `storage_for_inv`

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/util.rs:252]: hmm sorry I think I'm confused about this function now, so what's the difference between `storage_for` and `storage_for_inv`? Inverse of a value should be able to live in the same phase as the value, right?

File: Cargo.toml
Comment (ChihChengLiang) [inline Cargo.toml:17]: Do we intend to check in these changes?

```suggestion
halo2_proofs = { git = "https://github.com/privacy-scaling-explorations/halo2.git", tag = "v2022_09_10" }
```

Tip: Edu shared that we can patch in the root Cargo.toml.
```
[patch."https://github.com/privacy-scaling-explorations/halo2.git"]
halo2_proofs = { path = "../halo2/halo2_proofs" }
```

**Review by ChihChengLiang — COMMENTED** [review]
First pass

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/util.rs:252]: I understand, my question is more like, what makes inverse storage need one extra phase to be stored? For example `advice_at_phase_1 + challenge_usable_after_phase_1` we should be able to store its inverse in phase 2 right? Because this expression is evaluateable in phase 2.

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/util.rs:210]: Perhaps move to top?

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/util.rs:263]: Perhaps we could add comment that:

```rust
// The phase that given `Expression` becomes evaluateable.
```

File: zkevm-circuits/src/witness/block.rs
Comment (han0110) [inline zkevm-circuits/src/witness/block.rs:178]: Curious why does this become `0x10000`?

**Review by han0110 — COMMENTED** [review]
Finished my second review, only left some nitpicks. Also I think now it might be a good time to try to run integration test for this PR before being merged.

**Review by han0110 — APPROVED** [review]
Nice work!

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/execution/callop.rs:982]:
```suggestion
```

File: zkevm-circuits/src/evm_circuit/execution/extcodehash.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/execution/extcodehash.rs:66]: empty_hash

**Review by ChihChengLiang — APPROVED** [review]
LGTM. I added some nitpicks.

**Pull request merged**
